lib/three_scale/middleware/presigned_downloads.rb

```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+module ThreeScale
+  module Middleware
+    class PresignedDownloads
+      RE_SEPARATOR = ::File::SEPARATOR == "/" ? "/" : "[#{Regexp.escape(::File::SEPARATOR)}/]"
+
+      delegate *%i[requires_signing? good_signature? signature_at to_time], to: :class, private: true
+
+      class << self
+        def verifier
+          Rails.application.message_verifier("file downloads")
+        end
+
+        def requires_signing?(path)
+          # for the used separator see https://github.com/rack/rack/pull/2257
+          # question mark on first separator is paranoia of getting a relative path somehow
+          invoices_re = %r{#{RE_SEPARATOR}?system#{RE_SEPARATOR}.*#{RE_SEPARATOR}invoices#{RE_SEPARATOR}}
```

Comment:
If `RE_SEPARATOR` equals to `/`, then the expression would be:

```
/?system/.*/invoices/
```

This would be simpler and have the same effect:

```suggestion
          invoices_re = %r{system#{RE_SEPARATOR}.*#{RE_SEPARATOR}invoices#{RE_SEPARATOR}}
```

For the expression:

```
system/.*/invoices/
```

It would match the same paths, right?

Reply:
No, because we test with `start_with?`.

Reply:
Actually it should always start with `/` and seems like puma enforces this. But since Rack is accepting also without `/` , I am just being extra paranoid to catch the situation with a relative path. See the rack issue I linked to in description.